Feature: Object Pooling

mattpwest opened the pull request:
Object pooling implemented for:
 * Bullets
 * Paratroopers
 * Transports

Had to fix scoring as part of this and also attempted to address a bug where bullets often clipped through paratroopers.

mattpwest committed:
WIP on object pooling...

Started with bullets as the simplest case. Still have some logic bugs
causing shooting to only sometimes work and haven't started on
despawning logic yet...

mattpwest committed:
Merge branch 'master' into feature-object-pooling

mattpwest committed:
Pooling works properly for bullets!

mattpwest committed:
Switched to automatically creating pool parent prefabs instead of
passing them as Prefab configuration, since that wouldn't work for
prefabs that don't start in the scene...
Think I have paratroopers working properly as pooled objects (they were
quite complex due to parachutes).

mattpwest committed:
Transport planes are now also pooled!

mattpwest committed:
Attempted to fix Paratrooper hitbox (shots frequently clipped through).

sasvdw commented on ParatrooperUnity/Assets/Scripts/Damage.cs:
```diff
@@ -11,7 +11,21 @@ private void OnTriggerEnter2D(Collider2D other)
+
+	private void DestroySelf()
+	{
+		var pooled = gameObject.GetComponent<Pooled>();
```
use `RequiresComponent(typeof(Pooled))` attribute at the top and fetch the component in on Awake. It's a better form of dependency management, especially when behaviours need to be setup.

sasvdw commented on ParatrooperUnity/Assets/Scripts/Gun.cs:
```diff
@@ -38,8 +38,8 @@ public void Fire(Transform gunJoint)
 		var spawnPos = bulletSpawn.position;
-		var bullet = Instantiate(bulletPrefab, spawnPos, Quaternion.identity);
-		bullet.AddForce(direction * shotVelocity);
+		var bullet = PoolingFactory.SpawnOrRecycle(bulletPrefab.transform, spawnPos);
```
Add a generic method to for SpawnOrRecycle that can instantiate it and return the correct type by using the Instantiate<> method. That way the factory will return the object as the correct type.

sasvdw reviewed (changes requested):
Only one that I'm looking at here is to have the spawn or Recycle method be able to return the original type as well.

mattpwest committed:
Made suggested changes for review.
Attempted to apply same lessons learnt across the code base and
standardised formatting a bit (need to get your settings files, the one
you sent cannot be imported by Rider as far as I can see).

mattpwest committed:
That's what was stealing my typing in the diff window...
